app/src/main/java/tech/ula/utils/ServerUtility.kt

```diff
@@ -57,13 +55,19 @@ class ServerUtility(
     }
 
     /**
-     * Execute the boot script, should be called only when the session has the startOnBoot flag
-     * Users can use this to add their own custom commands to this script.
+     * Execute the user-defined startCommand
+     * Creates a new script based on userland_profile.sh and adds the user-defined command to the script
      */
-    private fun runBootScript(session: Session) {
+    fun executeStartCommand(session: Session) {
```

Comment:
I'd like to see at least one test for this function, mostly for regression purposes.

Reply:
You are going to have to help me on that because I never wrote a single unit test in my life.
What should we test for in this function?

Reply:
Sure, I can help you with that. Generally when writing unit tests, the minimum criteria is to write a test for each possible branch. So one test would just assert that if `executeStartCommand` is called with a `session` that had an empty `startCommand`, that it would return early. There are actually 2 ways to assert that for this case:

1. Assert that the `startScript` file is never overwritten.
2. Verify that `busyboxExecutor` is never called.

If you look in `ServerUtilityTest.kt` you can start seeing a bit about how these are tested. This should also start making it more clear why the dependencies that are injected into `ServerUtility` are injected into it: because it makes the class unit-testable. By provided mocks for the objects it uses, and a temporary folder to take the place of `applicationFilesDir`, we can guarantee that the tests will run as a unit anywhere we can run the JVM nd not just on the Android platform itself.

So onto writing the test itself:

```
    @Test
    fun `Calling executeStartCommand on a session without a start command returns early`() {
        // Given
        val filesystemId = 1L
        val session = Session(id = 0, filesystemId = filesystemId)
        val startScriptFile = File("${tempFolder.root}/$filesystemId/support/autostart.sh")
        val profileScriptFile = File("${tempFolder.root}/$filesystemId/support/userland_profile.sh")

        val startText = "start"
        startScriptFile.writeText(startText)

        val profileText = "profile"
        profileScriptFile.writeText(profileText)

        // When
        serverUtility.executeStartCommand(session)

        // Then
        val expectedCommand = "/support/autostart.sh"
        val expectedCommandShouldTerminate = false
        assertEquals(startText, startScriptFile.readText())
        verify(mockBusyboxExecutor, never()).executeProotCommand(
            eq(expectedCommand),
            eq(filesystemId.toString()),
            eq(expectedCommandShouldTerminate),
            anyOrNull(),
            anyOrNull()
        )
    }
```

Unit tests are typically written in a 'given, when, then` style. I've left some comments to show those stages for this test.

This test would pass, but in writing it it helped my understanding of some ways that the implementation of this function could be improved.

First of all, the function as it is written has an underlying (and hidden) assumption about a precondition. Namely, that `profileScript` already exists. That's a pretty safe assumption given the structure of the app as a whole, but it's worth accounting for it as a failure mode anyway. If the file doesn't exist, this function will throw an exception which isn't caught and handled explicitly elsewhere. That's probably a fine mode of failure in this case, but it leads nicely into another point: what happens if the command fails to execute? Right now, that would just result in a silent failure, which would likely lead to confusion on the part of the user. If the session is also going to start a server, that server and client would still be started. It would probably be better to stop starting the session and alert the user of the failure through a dialog. Since we would then have the infrastructure set up to alert of failures, maybe it's worth informing them of the failure case of the source `profileScript` not existing.

You can then write tests for the successful and failure cases pretty easily.

In terms of code structure, I think this function might actually be better placed in `BusyboxExecutor` itself. Nothing that this function does is explicitly related to servers. That being said, writing tests for it there would actually probably be harder. Feel free to leave it here for now, but maybe add a `TODO` comment to refactor it later.

If you're so inclined, [here](https://www.philosophicalhacker.com/2015/04/17/why-android-unit-testing-is-so-hard-pt-1/) is a series of blog posts that helped me deepen my understanding of the difficulties of unit testing on the Android platform.

If you have any more questions leave them here or find me in slack!

Reply:
Thank you for taking the time to explain this.
This is valuable information for me and I will take a look at the blog you linked.
The executeStartCommand function can definitely be improved.

We can add a few checks and I can think of a few ways to alert the user about the failure of execution :

1 - Send a broadcast from the service and have the MainActivity handle it and showing an error dialog. This would take care of sessions that are being started manually by the user, but would be useless on boot sessions because there is no activity listening.

2 - A new notification (or modifying the persistent one) to alert the user.

3 - We could send an intent to the MainActivity from the Service.
Something like :
```
Intent dialogIntent = new Intent(this, MyActivity.class);  
dialogIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);  
dialogIntent.putExtra("startBootFailed", true);
startActivity(dialogIntent);
```
When launching from a service, we need to use the FLAG_ACTIVITY_NEW_TASK.
This method works for both cases (user-initiated sessions and boot sessions) but needs to be handled differently in case the activity is already running. In this case, it will just be brought to the front without recreating the activity and will fire the onNewIntent() callback.

Reply:
I'm fine with either solution. Using a notification should probably create a new one rather than reuse the persistent so it's clear something happened. I imagine the notification would just lead to the app being opened anyway though, so maybe we just skip the middleman. Up to you.

Reply:
I will probably use option 3.
I have been busy with work but I plan on finishing the remaining issues this week.